src/base/bittorrent/torrenthandle.cpp

```diff
@@ -1744,45 +1749,64 @@ void TorrentHandle::handleFastResumeRejectedAlert(const lt::fastresume_rejected_
 
 void TorrentHandle::handleFileRenamedAlert(const lt::file_renamed_alert *p)
 {
-    const QString newName = Utils::Fs::fromNativePath(p->new_name());
-
-    // TODO: Check this!
-    if (filesCount() > 1) {
-        // Check if folders were renamed
-        QStringList oldPathParts = m_torrentInfo.origFilePath(p->index).split('/');
-        oldPathParts.removeLast();
-        QString oldPath = oldPathParts.join('/');
-        QStringList newPathParts = newName.split('/');
-        newPathParts.removeLast();
-        const QString newPath = newPathParts.join('/');
-        if (!newPathParts.isEmpty() && (oldPath != newPath)) {
-            qDebug("oldPath(%s) != newPath(%s)", qUtf8Printable(oldPath), qUtf8Printable(newPath));
-            oldPath = QString("%1/%2").arg(savePath(true), oldPath);
-            qDebug("Detected folder renaming, attempt to delete old folder: %s", qUtf8Printable(oldPath));
-            QDir().rmpath(oldPath);
-        }
-    }
-
     // We don't really need to call updateStatus() in this place.
     // All we need to do is make sure we have a valid instance of the TorrentInfo object.
     m_torrentInfo = TorrentInfo {m_nativeHandle.torrent_file()};
 
+    // remove empty leftover folders
+    const QString oldFilePath = m_oldPath[LTFileIndex {p->index}].takeFirst();
+    const QString newFilePath = Utils::Fs::fromNativePath(p->new_name());
+
+    if (m_oldPath[LTFileIndex {p->index}].isEmpty())
+        m_oldPath.remove(LTFileIndex {p->index});
+
+    QStringList oldPathParts = oldFilePath.split('/', QString::SkipEmptyParts);
+    oldPathParts.removeLast();  // drop file name part
+    QStringList newPathParts = newFilePath.split('/', QString::SkipEmptyParts);
+    newPathParts.removeLast();  // drop file name part
+
+#if defined(Q_OS_WIN)
+    const Qt::CaseSensitivity caseSensitivity = Qt::CaseInsensitive;
+#else
+    const Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive;
+#endif
+
+    int pathIdx = 0;
+    while ((pathIdx < oldPathParts.size()) && (pathIdx < newPathParts.size())) {
+        if (oldPathParts[pathIdx].compare(newPathParts[pathIdx], caseSensitivity) != 0)
+            break;
+        ++pathIdx;
+    }
+
+    for (int i = (oldPathParts.size() - 1); i >= pathIdx; --i) {
```

Comment:
Doesn't `Utils::Fs::smartRemoveEmptyFolderTree()` fit here?

Reply:
I think not in current form, `smartRemoveEmptyFolderTree()` will travel to *all* subdirectories which is not the behavior wanted here.
Although we can add another parameter to control it and also rename it: `smartRemoveEmptyFolder(const QString &path, const QDirIterator::IteratorFlags flags = QDirIterator::Subdirectories)`.

Reply:
On a second thought even using `QDirIterator::NoIteratorFlags` in `smartRemoveEmptyFolder`, it will iterate to every first level subfolder which is still wrong behavior for the situation here.
IMO let's keep the current way and change it when someone files a bug for it :P

Reply:
AFAIK, `smartRemoveEmptyFolderTree()` is designed to remove some folder that is logically empty but actually it can contain some unneeded files and other logically empty folders, isn't it? So why it "is not the behavior wanted here"?
E.g. initial layout:
```
FolderA/FolderB/FileA
FolderA/FolderB/FileB
FolderA/FolderC/FileC
```
Layout after you renamed (via qBittorrent GUI) FolderA to FolderN:
```
FolderA/FolderB/(empty)
FolderA/FolderC/(empty)
FolderN/FolderB/FileA
FolderN/FolderB/FileB
FolderN/FolderC/FileC
```
So you need "smart remove" FolderA with all its unneeded content. Or am I wrong?

Reply:
>So you need "smart remove" FolderA with all its unneeded content. Or am I wrong?

The files are being moved one by one, not in a batch.
So for each file move, you should only remove the affected (empty) folders.

For "FolderA/FolderB/FileA" -> "FolderN/FolderB/FileA", the folders to be deleted (if empty) are:
"FolderA/FolderB" & "FolderA".

Reply:
>Other sibling folders such as "FolderA/FolderC" should not be touched.

Ok. But how about "unneeded" files in direct parent folder if removed file? Is there no problem with it?

Reply:
>Is there no problem with it?

[As said](https://github.com/qbittorrent/qBittorrent/pull/10792#discussion_r292940315), personally I would like to revise it later if it poses a problem.